client/utils/guidance.py

```diff
@@ -129,49 +139,131 @@ def show_guidance_msg(self, unique_id, input_lines, access_token, hash_key, guid
+        log.info("Lambda Group: %s", lambda_string_key)
 
         lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
         if not lambda_info_misu:
             log.info("Cannot find info misU given the lambda string key.")
             print(GUIDANCE_DEFAULT_MSG)
             return EMPTY_MISUCOUNT_TGID_PRNTEDMSG
 
-        lst_mis_u = dict_info.get('lstMisU')
-
-        # No list of misunderstandings for this wrong answer, default message
-        if not lst_mis_u:
-            log.info("Cannot find the list of misunderstandings.")
+        shorten_unique_id = assess_id_util.canonicalize(unique_id)
+        # Try to get the info dictionary for this question. Maps wrong answer
+        # to dictionary
+        assess_dict_info = self.guidance_json[
+            'dictAssessId2Info'].get(shorten_unique_id)
+        if not assess_dict_info:
+            log.info("shorten_unique_id is not in dictAssessId2Info")
             print(GUIDANCE_DEFAULT_MSG)
             return EMPTY_MISUCOUNT_TGID_PRNTEDMSG
 
-        self.misU_count_dict = self.update_misUcounts(hash_key, lst_mis_u,
-                                                      repr(input_lines), shorten_unique_id)
+        wa_details = assess_dict_info['dictWA2DictInfo'].get(response)
+        if not wa_details:
+            log.info("Cannot find the wrong answer in the WA2Dict for this assesment.")
+            lst_mis_u = []
+        else:
+            lst_mis_u = wa_details.get('lstMisU', [])
 
-        Wrong_Answer_threshold = self.guidance_json['wrongAnsThresh']
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            log.info("Cannot find the list of misunderstandings.")
 
+        wa_count_threshold = self.guidance_json['wrongAnsThresh']
+        wa_lst_assess_num = assess_dict_info['dictWA2LstAssessNum_WA']
         msg_id_set = set()
 
-        # If the count is higher than the threshold, we need to display the
-        # message
-        for mis_u in self.misU_count_dict:
-            if self.misU_count_dict[mis_u] >= Wrong_Answer_threshold:
-                # Add each associated misunderstanding ID to the set
-                msg_id = lambda_info_misu(dict_info, mis_u)
+        answerDict, countData = self.get_misUdata()
+        prev_responses = answerDict.get(shorten_unique_id, [])
 
-                if msg_id is not None:
-                    msg_id_set.add(msg_id)
+        # Confirm that this WA has not been given before
+        seen_before = response in prev_responses
 
-        if self.tg_id == TG_CONTROL:
-            # if student is in control group, just print the default message
-            log.info("Student is in control group.")
-            print(GUIDANCE_DEFAULT_MSG)
-            return (self.misU_count_dict, self.tg_id, "")
+        answerDict[shorten_unique_id] = prev_responses + [response]
+        self.save_misUdata(answerDict, countData)
```

Comment:
These two lines should actually be in the `else` part of the `seen_before` if class. We don't want to save the answer at all if we've seen it before.